# Annexes

### Sequence attributes

#### Reserved sequence attributes

##### `ali_dir`

###### Type : `string`

The attribute can contain 2 string values `"left"` or `"right".`

###### Set by the *obipairing* tool

The alignment generated by *obipairing* is a 3'-end gap free algorithm.
Two cases can occur when aligning the forward and reverse reads. If the
barcode is long enough, both the reads overlap only on their 3' ends. In
such case, the alignment direction `ali_dir` is set to *left*. If the
barcode is shorter than the read length, the paired reads overlap by
their 5' ends, and the complete barcode is sequenced by both the reads.
In that later case, `ali_dir` is set to *right*.

##### `ali_length`

###### Set by the *obipairing* tool

Length of the aligned parts when merging forward and reverse reads

##### `count` : the number of sequence occurrences

###### Set by the *obiuniq* tool

The `count` attribute indicates how-many strictly identical sequences
have been merged in a single record. It contains an integer value. If it
is absent this means that the sequence record represents a single
occurrence of the sequence.

###### Getter : method `Count()`

The `Count()` method allows to access to the count attribute as an
integer value. If the `count` attribute is not defined for the given
sequence, the value *1* is returned

##### `merged_*`

###### Type : `map[string]int`

###### Set by the *obiuniq* tool

The `-m` option of the *obiuniq* tools allows for keeping track of the
distribution of the values stored in given attribute of interest. Often
this option is used to summarise distribution of a sequence variant
accross samples when *obiuniq* is run after running *obimultiplex*. The
actual name of the attribute depends on the name of the monitored
attribute. If `-m` option is used with the attribute *sample*, then this
attribute names *merged_sample*.

##### `mode`

###### Set by the *obipairing* tool

**`obitag_ref_index`**

###### Set by the *obirefidx* tool.

It resumes to which taxonomic annotation a match to that sequence must
lead according to the number of differences existing between the query
sequence and the reference sequence having that tag.

###### Getter : method `Count()`

##### `pairing_mismatches`

###### Set by the *obipairing* tool

##### `score`

###### Set by the *obipairing* tool

##### `score_norm`

###### Set by the *obipairing* tool
